components/service/sync-logins/src/test/java/mozilla/components/service/sync/logins/AsyncLoginsStorageAdapterTest.kt

```diff
@@ -0,0 +1,215 @@
+package mozilla.components.service.sync.logins
+
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withTimeoutOrNull
+import mozilla.appservices.logins.LoginsStorage
+import mozilla.appservices.logins.MismatchedLockException
+import mozilla.appservices.logins.ServerPassword
+import mozilla.appservices.logins.SyncUnlockInfo
+import mozilla.appservices.sync15.SyncTelemetryPing
+import org.json.JSONObject
+import org.junit.Before
+import org.junit.Test
+import org.mockito.Mockito.verify
+import org.mockito.Mockito.spy
+import org.mockito.Mockito.times
+
+@ExperimentalCoroutinesApi
+class AsyncLoginsStorageAdapterTest {
+
+    private val key = { GlobalScope.async { "key" } }
+    private val exceptionHandler = CoroutineExceptionHandler { _, e ->
+        runBlocking {
+            exceptionChannel.send(e)
+        }
+    }
+
+    private lateinit var exceptionChannel: Channel<Throwable>
+    private lateinit var storage: LoginsStorage
+    private lateinit var adapter: AsyncLoginsStorageAdapter<LoginsStorage>
+
+    @Before
+    fun before() {
+        exceptionChannel = Channel()
+        storage = spy(TestLoginsStorage())
+        adapter = AsyncLoginsStorageAdapter(storage)
+    }
+
+    @Suppress("DeferredResultUnused")
+    @Test(expected = LoginsStorageException::class)
+    fun `unsafe locking and unlocking should throw exception`() {
+        GlobalScope.launch(exceptionHandler) {
+            with(adapter) {
+                unlock(key().await())
+                delay(10)
+                list().await()
+                lock()
+            }
+        }
+
+        GlobalScope.launch(exceptionHandler) {
+            with(adapter) {
+                unlock(key().await())
+                list().await()
+                lock()
+            }
+        }
+
+        runBlocking {
+            withTimeoutOrNull(100) {
+                for (e in exceptionChannel) {
+                    throw e
```

Comment:
This was the best way I could come up with to reproduce this in a test.  I'm open to suggestions.